File: src/lib/renderers/visual/WebGPUBrush.ts
import wgslRequires from './wgsl/wgslRequires.wgsl?raw'
import { sumBy } from 'lodash'
import calcPosition from './wgsl/calcPosition'
import { Scene } from '@/lib/types'

export type AsemicGroup = {
  points: any[][]
  settings: any
}

export default abstract class WebGPUBrush {
  ctx: GPUCanvasContext
  device: GPUDevice
  pipeline: GPURenderPipeline
  bindGroup: GPUBindGroup
  bindGroup2: GPUBindGroup
  shaderModule: GPUShaderModule
  dimensions: { buffer: GPUBuffer; size: number }
  time: { buffer: GPUBuffer; size: number }
  scrub: { buffer: GPUBuffer; size: number }
  widths: { buffer: GPUBuffer; size: number }
  curveStarts: { buffer: GPUBuffer; size: number; array: Uint32Array }
  vertex: { buffer: GPUBuffer; size: number }
  index: { buffer: GPUBuffer; size: number }
  colors: { buffer: GPUBuffer; size: number }
  texture: {
    src: GPUTexture
    imageData: ImageData[]
    transformBuffer: GPUBuffer
  } | null = null
  settings: any

  abstract get mode(): string

  destroy() {
    this.vertex?.buffer.destroy()
    this.index?.buffer.destroy()
    this.dimensions?.buffer.destroy()
    this.time?.buffer.destroy()
    this.widths?.buffer.destroy()
    this.curveStarts?.buffer.destroy()
    this.colors?.buffer.destroy()
    this.texture?.src.destroy()
    this.texture?.transformBuffer.destroy()
  }

  protected abstract loadIndex(curves: AsemicGroup): Uint32Array
  protected abstract loadPipeline(
    bindGroupLayout: GPUBindGroupLayout
  ): GPURenderPipeline
  protected loadShader({ includeTexture = false } = {}) {
    return this.device.createShaderModule({
      code: /*wgsl*/ `
    struct VertexOutput {
    @builtin(position) position: vec4<f32>,
    @location(0) color: vec4<f32>,
    @location(1) uv: vec2<f32>,
    @location(2) tangent: vec2<f32>,
    @location(3) t: f32,
    };
    
    @group(0) @binding(0)
    var<storage, read> vertices: array<vec2<f32>>;

    @group(0) @binding(1)
    var<storage, read> widths: array<f32>;

    @group(0) @binding(2)
    var<uniform> canvas_dimensions: vec2<f32>;

    @group(0) @binding(3)
    var<storage, read> curve_starts: array<u32>;

    @group(0) @binding(4)
    var<storage, read> colors: array<vec4<f32>>;

    @group(0) @binding(5)
    var<uniform> time: f32;

    @group(0) @binding(6)
    var<uniform> scrub: f32;

    ${
      includeTexture
        ? /*wgsl*/ `@group(1) @binding(0) var textureSampler: sampler;
      @group(1) @binding(1) var tex: texture_2d<f32>;
      @group(1) @binding(2)
      var<uniform> texture_transform: vec4<f32>; // xy offset, wh scale`
        : ''
    }

    ${wgslRequires}

    @vertex
    fn vertexMain(@builtin(vertex_index) vertex_index: u32) -> VertexOutput {
    var output: VertexOutput;
    ${calcPosition(this)}
    return output;
    }

    @fragment
    fn fragmentMain(input: VertexOutput) -> @location(0) vec4<f32> {
    var texColor = input.color;
    // Sample texture if present, otherwise use white
    ${
      includeTexture
        ? /*wgsl*/ `
        // Get screen-space UV coordinates
        let screenUV = input.position.xy / canvas_dimensions;
        
        // Apply texture offset and scale transformation
        // offset.xy is in 0-1 range, scale it to canvas dimensions
        let offsetPixels = texture_transform.xy * canvas_dimensions;
        let scaleFactors = texture_transform.zw;
        
        // Transform UV coordinates: offset and scale
        let transformedUV = (screenUV * canvas_dimensions - offsetPixels) / (scaleFactors * canvas_dimensions);
        
        // Sample with wrapping (fract provides the wrapping behavior)
        texColor = texColor * textureSample(tex, textureSampler, fract(transformedUV));`
        : ''
    }
    return texColor;
    }
    `
    })
  }

  protected reload(curves: AsemicGroup) {
    // const vertices = new Float32Array(
    //   curves.flatMap(x => x.flatMap(x => [x.x, x.y]))
    // )
    const vertices = new Float32Array(
      curves.points.flatMap(x => x.flatMap(x => [x.x, x.y]))
    )

    this.device.queue.writeBuffer(this.vertex.buffer, 0, vertices)

    const widths = new Float32Array(
      curves.points.flatMap(x => x.flatMap(x => x.w))
    )
    this.device.queue.writeBuffer(this.widths.buffer, 0, widths)

    const colors = new Float32Array(
      curves.points.flatMap(x => x.flatMap(x => [x.h, x.s, x.l, x.a]))
    )

    this.device.queue.writeBuffer(this.colors.buffer, 0, colors)

    // Create a buffer for canvas dimensions
    const canvasDimensions = new Float32Array([
      this.ctx.canvas.width,
      this.ctx.canvas.height
    ])
    this.device.queue.writeBuffer(this.dimensions.buffer, 0, canvasDimensions)

    if (this.texture) {
      const texture = this.texture.src
      const imageData = this.texture.imageData
      this.device.queue.writeTexture(
        { texture },
        imageData[0].data,
        { bytesPerRow: imageData[0].width * 4 },
        [imageData[0].width, imageData[0].height]
      )
      const xy = curves.xy
      const wh = curves.wh
      const textureTransform = new Float32Array([xy[0], xy[1], wh[0], wh[1]])
      this.device.queue.writeBuffer(
        this.texture.transformBuffer,
        0,
        textureTransform
      )
    }
  }

  load(group: AsemicGroup) {
    const vertexBuffer = this.device.createBuffer({
      size:
        Float32Array.BYTES_PER_ELEMENT * sumBy(group.points, x => x.length * 2),
      usage: GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_DST,
      mappedAtCreation: false,
      label: 'vertex'
    })

    // Create an index buffer
    const widthsBuffer = this.device.createBuffer({
      size:
        Float32Array.BYTES_PER_ELEMENT * sumBy(group.points, x => x.length * 2),
      usage: GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_DST,
      mappedAtCreation: false,
      label: 'widths'
    })

    const colorsBuffer = this.device.createBuffer({
      size:
        Float32Array.BYTES_PER_ELEMENT * sumBy(group.points, x => x.length * 4),
      usage: GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_DST,
      mappedAtCreation: false,
      label: 'colors'
    })

    const indices = this.loadIndex(group)

    // Create an index buffer
    const indexBuffer = this.device.createBuffer({
      size: indices.byteLength,
      usage: GPUBufferUsage.INDEX,
      mappedAtCreation: true,
      label: 'index buffer'
    })

    // Write index data
    new Uint32Array(indexBuffer.getMappedRange()).set(indices)
    indexBuffer.unmap()

    // Define curve start indices
    const curveStarts = new Uint32Array(group.points.length + 1)
    let total = 0
    for (let i = 0; i < group.points.length; i++) {
      curveStarts[i] = total
      total += group.points[i].length
    }
    curveStarts[group.points.length] = total

    // Create a buffer for curve starts
    const curveStartsBuffer = this.device.createBuffer({
      size: curveStarts.byteLength,
      usage: GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_DST,
      mappedAtCreation: true,
      label: 'curve starts'
    })

    // Write curve starts data
    new Uint32Array(curveStartsBuffer.getMappedRange()).set(curveStarts)
    curveStartsBuffer.unmap()

    const dimensionsBuffer = this.device.createBuffer({
      size: Float32Array.BYTES_PER_ELEMENT * 2,
      usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
      mappedAtCreation: false,
      label: 'canvas dimensions'
    })

    const timeBuffer = this.device.createBuffer({
      size: Float32Array.BYTES_PER_ELEMENT * 1,
      usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
      mappedAtCreation: false,
      label: 'time'
    })

    const scrubBuffer = this.device.createBuffer({
      size: Float32Array.BYTES_PER_ELEMENT * 1,
      usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
      mappedAtCreation: false,
      label: 'scrub'
    })

    const bindGroupLayoutEntries: Array<GPUBindGroupLayoutEntry> = [
      {
        binding: 0,
        visibility: GPUShaderStage.VERTEX,
        buffer: { type: 'read-only-storage' }
      },
      {
        binding: 1,
        visibility: GPUShaderStage.VERTEX,
        buffer: { type: 'read-only-storage' }
      },
      {
        binding: 2,
        visibility: GPUShaderStage.VERTEX | GPUShaderStage.FRAGMENT,
        buffer: { type: 'uniform' }
      },
      {
        binding: 3,
        visibility: GPUShaderStage.VERTEX,
        buffer: { type: 'read-only-storage' }
      },
      {
        binding: 4,
        visibility: GPUShaderStage.VERTEX,
        buffer: { type: 'read-only-storage' }
      },
      {
        binding: 5,
        visibility: GPUShaderStage.VERTEX | GPUShaderStage.FRAGMENT,
        buffer: { type: 'uniform' }
      },
      {
        binding: 6,
        visibility: GPUShaderStage.VERTEX | GPUShaderStage.FRAGMENT,
        buffer: { type: 'uniform' }
      }
    ]
    const bindGroupEntries: Array<GPUBindGroupEntry> = [
      {
        binding: 0,
        resource: { buffer: vertexBuffer }
      },
      {
        binding: 1,
        resource: { buffer: widthsBuffer }
      },
      {
        binding: 2,
        resource: { buffer: dimensionsBuffer }
      },
      {
        binding: 3,
        resource: { buffer: curveStartsBuffer }
      },
      {
        binding: 4,
        resource: { buffer: colorsBuffer }
      },
      {
        binding: 5,
        resource: { buffer: timeBuffer }
      },
      {
        binding: 6,
        resource: { buffer: scrubBuffer }
      }
    ]

    if (group.settings.texture && group.imageDatas) {
      const imageData = group.imageDatas[0] as ImageData
      const texture = this.device.createTexture({
        size: [imageData.width, imageData.height, 1],
        format: 'rgba8unorm',
        usage:
          GPUTextureUsage.TEXTURE_BINDING |
          GPUTextureUsage.COPY_DST |
          GPUTextureUsage.RENDER_ATTACHMENT
      })

      // Create texture transform buffer
      const textureTransformBuffer = this.device.createBuffer({
        size: Float32Array.BYTES_PER_ELEMENT * 4, // xy offset + wh scale
        usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
        mappedAtCreation: false,
        label: 'texture transform'
      })

      const bindGroup2 = this.device.createBindGroup({
        layout: this.device.createBindGroupLayout({
          entries: [
            {
              binding: 6,
              visibility: GPUShaderStage.FRAGMENT,
              sampler: {}
            },
            {
              binding: 7,
              visibility: GPUShaderStage.FRAGMENT,
              texture: {}
            },
            {
              binding: 8,
              visibility: GPUShaderStage.FRAGMENT,
              buffer: { type: 'uniform' }
            }
          ]
        }),
        entries: [
          {
            binding: 6,
            resource: this.device.createSampler({
              addressModeU: 'repeat',
              addressModeV: 'repeat',
              magFilter: 'linear',
              minFilter: 'linear'
            })
          },
          {
            binding: 7,
            resource: texture.createView()
          },
          {
            binding: 8,
            resource: { buffer: textureTransformBuffer }
          }
        ]
      })
      this.bindGroup2 = bindGroup2

      // Set texture transform data (xy offset, wh scale)
      this.texture = {
        src: texture,
        transformBuffer: textureTransformBuffer,
        imageData: [imageData]
      }
    } else {
      this.texture = null
    }

    // Update bind group layout to include dimensions uniform
    const bindGroupLayout = this.device.createBindGroupLayout({
      entries: bindGroupLayoutEntries
    })
    const bindGroup = this.device.createBindGroup({
      layout: bindGroupLayout,
      entries: bindGroupEntries
    })

    // Texture support: add bindings for texture and sampler if textures are available
    const pipeline = this.loadPipeline(bindGroupLayout)
    // Store pipeline and resources as instance properties
    this.pipeline = pipeline
    this.bindGroup = bindGroup

    if (this.index) {
      this.index.buffer.destroy()
      this.dimensions.buffer.destroy()
      this.widths.buffer.destroy()
      this.curveStarts.buffer.destroy()
      this.vertex.buffer.destroy()
      this.colors.buffer.destroy()
    }
    this.colors = {
      buffer: colorsBuffer,
      size: sumBy(group.points, x => x.length * 4)
    }

    this.index = { buffer: indexBuffer, size: indices.length }

    this.dimensions = { buffer: dimensionsBuffer, size: 2 }
    this.time = { buffer: timeBuffer, size: 1 }
    this.scrub = { buffer: scrubBuffer, size: 1 }
    this.widths = {
      buffer: widthsBuffer,
      size: sumBy(group.points, x => x.length * 2)
    }
    this.curveStarts = {
      buffer: curveStartsBuffer,
      size: curveStarts.length,
      array: curveStarts
    }
    this.vertex = {
      buffer: vertexBuffer,
      size: sumBy(group.points, x => x.length * 2)
    }
  }

  render(curves: AsemicGroup, renderPass: GPURenderPassEncoder, scene: Scene) {
    if (
      curves.points.length === 0 ||
      (curves.points.length < 2 && curves.points[0].length < 2)
    ) {
      return
    }
    // debugger

    if (!this.vertex) {
      this.load(curves)
    } else if (
      !this.texture &&
      curves.imageDatas &&
      curves.imageDatas[0].data.find(x => x > 0)
    ) {
      this.shaderModule = this.loadShader({ includeTexture: true })
      this.load(curves)
    } else if (this.curveStarts.size !== curves.points.length + 1) {
      this.load(curves)
    } else {
      let total = 0
      for (let i = 0; i < curves.points.length + 1; i++) {
        if (this.curveStarts.array[i] !== total) {
          this.load(curves)
          break
        }
        if (!curves.points[i]) break
        total += curves.points[i].length
      }
    }
    this.reload(curves)

    this.device.queue.writeBuffer(
      this.time.buffer,
      0,
      new Float32Array([performance.now() / 1000])
    )

    this.device.queue.writeBuffer(
      this.scrub.buffer,
      0,
      new Float32Array([scene.scrub])
    )

    // These operations could all be done once during init
    renderPass.setBindGroup(0, this.bindGroup)
    renderPass.setPipeline(this.pipeline)
    renderPass.setIndexBuffer(this.index.buffer, 'uint32')
    // This needs to stay in the render method
    renderPass.drawIndexed(this.index.size)
  }

  constructor(
    ctx: GPUCanvasContext,
    device: GPUDevice,
    settings: AsemicGroup['settings']
  ) {
    this.ctx = ctx
    this.device = device
    this.settings = settings
    this.shaderModule = this.loadShader()
  }
}

File: src/lib/renderers/visual/webgpubrush.rs
use crate::lib::types::Scene;
use std::num::NonZeroU64;
use wgpu::{
    BindGroup, BindGroupLayout, Buffer, BufferUsages, CanvasContext, Device, Queue,
    RenderPassEncoder, RenderPipeline, ShaderModule, Texture, TextureSampler, TextureUsages,
};

/// Represents a group of curves with associated settings
#[derive(Clone, Debug)]
pub struct AsemicGroup {
    pub points: Vec<Vec<Point>>,
    pub settings: CurveSettings,
}

/// A single point in a curve with rendering attributes
#[derive(Clone, Debug)]
pub struct Point {
    pub x: f32,
    pub y: f32,
    pub w: f32,          // width
    pub h: f32,          // hue
    pub s: f32,          // saturation
    pub l: f32,          // lightness
    pub a: f32,          // alpha
    pub attrs: [f32; 4], // additional attributes
}

/// Settings for curve rendering
#[derive(Clone, Debug, Default)]
pub struct CurveSettings {
    // Curve-specific settings can be added here
}

/// Texture support for brush rendering
pub struct BrushTexture {
    pub src: Texture,
    pub image_data: Vec<wgpu::ImageCopyTexture<'static>>,
    pub transform_buffer: Buffer,
}

/// Abstract trait for WebGPU brush implementations
pub trait WebGPUBrushImpl: Send + Sync {
    fn mode(&self) -> &str;
    fn load_index(&self, curves: &AsemicGroup) -> Vec<u32>;
    fn load_pipeline(&self, device: &Device, bind_group_layout: &BindGroupLayout)
        -> RenderPipeline;
}

/// WebGPU Brush for rendering curves
pub struct WebGPUBrush {
    pub ctx: CanvasContext,
    pub device: Device,
    pub queue: Queue,
    pub pipeline: Option<RenderPipeline>,
    pub bind_group: Option<BindGroup>,
    pub shader_module: ShaderModule,

    // GPU Buffers
    pub dimensions: BufferBinding,
    pub time: BufferBinding,
    pub scrub: BufferBinding,
    pub widths: BufferBinding,
    pub curve_starts: CurveStartsBinding,
    pub vertex: BufferBinding,
    pub index: IndexBinding,
    pub colors: BufferBinding,
    pub attrs: BufferBinding,

    // Optional texture
    pub texture: Option<BrushTexture>,

    // Settings
    pub settings: CurveSettings,

    // Implementation
    pub impl_: Box<dyn WebGPUBrushImpl>,
}

/// Wrapper for buffer bindings with size tracking
#[derive(Clone)]
pub struct BufferBinding {
    pub buffer: Buffer,
    pub size: usize,
}

/// Wrapper for index buffer with size
pub struct IndexBinding {
    pub buffer: Buffer,
    pub size: usize,
}

/// Wrapper for curve starts with array copy
pub struct CurveStartsBinding {
    pub buffer: Buffer,
    pub size: usize,
    pub array: Vec<u32>,
}

impl Drop for WebGPUBrush {
    fn drop(&mut self) {
        // GPU resources are automatically cleaned up by wgpu
        // No explicit cleanup needed in Rust
    }
}

impl WebGPUBrush {
    /// Load shader module with optional texture support
    pub fn load_shader(&self, include_texture: bool) -> ShaderModule {
        let shader_code = if include_texture {
            include_str!("./wgsl/brush_shader_with_texture.wgsl")
        } else {
            include_str!("./wgsl/brush_shader.wgsl")
        };

        self.device
            .create_shader_module(wgpu::ShaderModuleDescriptor {
                label: Some("brush_shader"),
                source: wgpu::ShaderSource::Wgsl(shader_code.into()),
            })
    }

    /// Reload buffer data from updated curves
    pub fn reload(&mut self, curves: &AsemicGroup) {
        // Prepare vertex data
        let vertices: Vec<f32> = curves
            .points
            .iter()
            .flat_map(|curve| curve.iter().flat_map(|point| vec![point.x, point.y]))
            .collect();

        self.queue
            .write_buffer(&self.vertex.buffer, 0, bytemuck::cast_slice(&vertices));

        // Prepare width data
        let widths: Vec<f32> = curves
            .points
            .iter()
            .flat_map(|curve| curve.iter().map(|point| point.w))
            .collect();

        self.queue
            .write_buffer(&self.widths.buffer, 0, bytemuck::cast_slice(&widths));

        // Prepare color data (HSLA)
        let colors: Vec<f32> = curves
            .points
            .iter()
            .flat_map(|curve| {
                curve
                    .iter()
                    .flat_map(|point| vec![point.h, point.s, point.l, point.a])
            })
            .collect();

        self.queue
            .write_buffer(&self.colors.buffer, 0, bytemuck::cast_slice(&colors));

        // Prepare attributes data
        let attrs: Vec<f32> = curves
            .points
            .iter()
            .flat_map(|curve| curve.iter().flat_map(|point| point.attrs.to_vec()))
            .collect();

        self.queue
            .write_buffer(&self.attrs.buffer, 0, bytemuck::cast_slice(&attrs));

        // Update canvas dimensions
        let canvas_dims = [self.ctx.canvas.width as f32, self.ctx.canvas.height as f32];
        self.queue.write_buffer(
            &self.dimensions.buffer,
            0,
            bytemuck::cast_slice(&canvas_dims),
        );
    }

    /// Load curve data into GPU buffers
    pub fn load(&mut self, group: &AsemicGroup) {
        println!("loading curves");

        // Calculate total vertex count
        let total_vertex_count: usize = group.points.iter().map(|curve| curve.len()).sum();

        // Create vertex buffer
        let vertex_buffer = self.device.create_buffer(&wgpu::BufferDescriptor {
            label: Some("vertex_buffer"),
            size: (total_vertex_count * 2 * std::mem::size_of::<f32>()) as u64,
            usage: BufferUsages::STORAGE | BufferUsages::COPY_DST,
            mapped_at_creation: false,
        });

        // Create widths buffer
        let widths_buffer = self.device.create_buffer(&wgpu::BufferDescriptor {
            label: Some("widths_buffer"),
            size: (total_vertex_count * std::mem::size_of::<f32>()) as u64,
            usage: BufferUsages::STORAGE | BufferUsages::COPY_DST,
            mapped_at_creation: false,
        });

        // Create colors buffer (HSLA per vertex)
        let colors_buffer = self.device.create_buffer(&wgpu::BufferDescriptor {
            label: Some("colors_buffer"),
            size: (total_vertex_count * 4 * std::mem::size_of::<f32>()) as u64,
            usage: BufferUsages::STORAGE | BufferUsages::COPY_DST,
            mapped_at_creation: false,
        });

        // Create attributes buffer
        let attrs_buffer = self.device.create_buffer(&wgpu::BufferDescriptor {
            label: Some("attrs_buffer"),
            size: (total_vertex_count * 4 * std::mem::size_of::<f32>()) as u64,
            usage: BufferUsages::STORAGE | BufferUsages::COPY_DST,
            mapped_at_creation: false,
        });

        // Load indices from implementation
        let indices = self.impl_.load_index(group);

        // Create index buffer
        let index_buffer = self
            .device
            .create_buffer_init(&wgpu::util::BufferInitDescriptor {
                label: Some("index_buffer"),
                contents: bytemuck::cast_slice(&indices),
                usage: BufferUsages::INDEX | BufferUsages::COPY_DST,
            });

        // Calculate curve starts
        let mut curve_starts = Vec::with_capacity(group.points.len() + 1);
        let mut total = 0u32;
        for curve in &group.points {
            curve_starts.push(total);
            total += curve.len() as u32;
        }
        curve_starts.push(total);

        // Create curve starts buffer
        let curve_starts_buffer =
            self.device
                .create_buffer_init(&wgpu::util::BufferInitDescriptor {
                    label: Some("curve_starts_buffer"),
                    contents: bytemuck::cast_slice(&curve_starts),
                    usage: BufferUsages::STORAGE | BufferUsages::COPY_DST,
                });

        // Create dimensions buffer
        let dimensions_buffer = self.device.create_buffer(&wgpu::BufferDescriptor {
            label: Some("dimensions_buffer"),
            size: (2 * std::mem::size_of::<f32>()) as u64,
            usage: BufferUsages::UNIFORM | BufferUsages::COPY_DST,
            mapped_at_creation: false,
        });

        // Create time buffer
        let time_buffer = self.device.create_buffer(&wgpu::BufferDescriptor {
            label: Some("time_buffer"),
            size: std::mem::size_of::<f32>() as u64,
            usage: BufferUsages::UNIFORM | BufferUsages::COPY_DST,
            mapped_at_creation: false,
        });

        // Create scrub buffer
        let scrub_buffer = self.device.create_buffer(&wgpu::BufferDescriptor {
            label: Some("scrub_buffer"),
            size: (2 * std::mem::size_of::<f32>()) as u64,
            usage: BufferUsages::UNIFORM | BufferUsages::COPY_DST,
            mapped_at_creation: false,
        });

        // Create bind group layout
        let bind_group_layout =
            self.device
                .create_bind_group_layout(&wgpu::BindGroupLayoutDescriptor {
                    label: Some("brush_bind_group_layout"),
                    entries: &[
                        // Vertices (binding 0)
                        wgpu::BindGroupLayoutEntry {
                            binding: 0,
                            visibility: wgpu::ShaderStages::VERTEX,
                            ty: wgpu::BindingType::Buffer {
                                ty: wgpu::BufferBindingType::Storage { read_only: true },
                                has_dynamic_offset: false,
                                min_binding_size: None,
                            },
                            count: None,
                        },
                        // Widths (binding 1)
                        wgpu::BindGroupLayoutEntry {
                            binding: 1,
                            visibility: wgpu::ShaderStages::VERTEX,
                            ty: wgpu::BindingType::Buffer {
                                ty: wgpu::BufferBindingType::Storage { read_only: true },
                                has_dynamic_offset: false,
                                min_binding_size: None,
                            },
                            count: None,
                        },
                        // Dimensions (binding 2)
                        wgpu::BindGroupLayoutEntry {
                            binding: 2,
                            visibility: wgpu::ShaderStages::VERTEX | wgpu::ShaderStages::FRAGMENT,
                            ty: wgpu::BindingType::Buffer {
                                ty: wgpu::BufferBindingType::Uniform,
                                has_dynamic_offset: false,
                                min_binding_size: None,
                            },
                            count: None,
                        },
                        // Curve starts (binding 3)
                        wgpu::BindGroupLayoutEntry {
                            binding: 3,
                            visibility: wgpu::ShaderStages::VERTEX,
                            ty: wgpu::BindingType::Buffer {
                                ty: wgpu::BufferBindingType::Storage { read_only: true },
                                has_dynamic_offset: false,
                                min_binding_size: None,
                            },
                            count: None,
                        },
                        // Colors (binding 4)
                        wgpu::BindGroupLayoutEntry {
                            binding: 4,
                            visibility: wgpu::ShaderStages::VERTEX,
                            ty: wgpu::BindingType::Buffer {
                                ty: wgpu::BufferBindingType::Storage { read_only: true },
                                has_dynamic_offset: false,
                                min_binding_size: None,
                            },
                            count: None,
                        },
                        // Time (binding 5)
                        wgpu::BindGroupLayoutEntry {
                            binding: 5,
                            visibility: wgpu::ShaderStages::VERTEX | wgpu::ShaderStages::FRAGMENT,
                            ty: wgpu::BindingType::Buffer {
                                ty: wgpu::BufferBindingType::Uniform,
                                has_dynamic_offset: false,
                                min_binding_size: None,
                            },
                            count: None,
                        },
                        // Scrub (binding 6)
                        wgpu::BindGroupLayoutEntry {
                            binding: 6,
                            visibility: wgpu::ShaderStages::VERTEX | wgpu::ShaderStages::FRAGMENT,
                            ty: wgpu::BindingType::Buffer {
                                ty: wgpu::BufferBindingType::Uniform,
                                has_dynamic_offset: false,
                                min_binding_size: None,
                            },
                            count: None,
                        },
                        // Attributes (binding 7)
                        wgpu::BindGroupLayoutEntry {
                            binding: 7,
                            visibility: wgpu::ShaderStages::VERTEX,
                            ty: wgpu::BindingType::Buffer {
                                ty: wgpu::BufferBindingType::Storage { read_only: true },
                                has_dynamic_offset: false,
                                min_binding_size: None,
                            },
                            count: None,
                        },
                    ],
                });

        // Create bind group
        let bind_group = self.device.create_bind_group(&wgpu::BindGroupDescriptor {
            label: Some("brush_bind_group"),
            layout: &bind_group_layout,
            entries: &[
                wgpu::BindGroupEntry {
                    binding: 0,
                    resource: vertex_buffer.as_entire_binding(),
                },
                wgpu::BindGroupEntry {
                    binding: 1,
                    resource: widths_buffer.as_entire_binding(),
                },
                wgpu::BindGroupEntry {
                    binding: 2,
                    resource: dimensions_buffer.as_entire_binding(),
                },
                wgpu::BindGroupEntry {
                    binding: 3,
                    resource: curve_starts_buffer.as_entire_binding(),
                },
                wgpu::BindGroupEntry {
                    binding: 4,
                    resource: colors_buffer.as_entire_binding(),
                },
                wgpu::BindGroupEntry {
                    binding: 5,
                    resource: time_buffer.as_entire_binding(),
                },
                wgpu::BindGroupEntry {
                    binding: 6,
                    resource: scrub_buffer.as_entire_binding(),
                },
                wgpu::BindGroupEntry {
                    binding: 7,
                    resource: attrs_buffer.as_entire_binding(),
                },
            ],
        });

        // Load pipeline from implementation
        let pipeline = self.impl_.load_pipeline(&self.device, &bind_group_layout);

        // Clean up old resources
        if let Some(old_index) = self.index.take() {
            // wgpu handles cleanup automatically
        }

        // Update instance state
        self.colors = BufferBinding {
            buffer: colors_buffer,
            size: total_vertex_count * 4,
        };
        self.attrs = BufferBinding {
            buffer: attrs_buffer,
            size: total_vertex_count * 4,
        };
        self.index = Some(IndexBinding {
            buffer: index_buffer,
            size: indices.len(),
        });
        self.dimensions = BufferBinding {
            buffer: dimensions_buffer,
            size: 2,
        };
        self.time = BufferBinding {
            buffer: time_buffer,
            size: 1,
        };
        self.scrub = BufferBinding {
            buffer: scrub_buffer,
            size: 2,
        };
        self.widths = BufferBinding {
            buffer: widths_buffer,
            size: total_vertex_count,
        };
        self.curve_starts = CurveStartsBinding {
            buffer: curve_starts_buffer,
            size: curve_starts.len(),
            array: curve_starts,
        };
        self.vertex = BufferBinding {
            buffer: vertex_buffer,
            size: total_vertex_count * 2,
        };

        self.pipeline = Some(pipeline);
        self.bind_group = Some(bind_group);
    }

    /// Render the curves
    pub fn render(
        &mut self,
        curves: &AsemicGroup,
        render_pass: &mut RenderPassEncoder,
        scene: &Scene,
    ) {
        // Early exit if no curves
        if curves.points.is_empty() || (curves.points.len() < 2 && curves.points[0].len() < 2) {
            return;
        }

        // Load or reload buffers if needed
        if self.vertex.buffer.size() == 0 {
            self.load(curves);
        } else if self.curve_starts.size != curves.points.len() + 1 {
            self.load(curves);
        } else {
            // Check if curve sizes have changed
            let mut total = 0u32;
            let mut needs_reload = false;
            for (i, curve) in curves.points.iter().enumerate() {
                if self.curve_starts.array[i] != total {
                    needs_reload = true;
                    break;
                }
                total += curve.len() as u32;
            }
            if needs_reload {
                self.load(curves);
            }
        }

        self.reload(curves);

        // Update time uniform
        let time = (std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_secs_f64()
            * 1000.0) as f32
            / 1000.0;
        self.queue
            .write_buffer(&self.time.buffer, 0, bytemuck::cast_slice(&[time]));

        // Update scrub uniform
        self.queue.write_buffer(
            &self.scrub.buffer,
            0,
            bytemuck::cast_slice(&[scene.scrub, scene.fade]),
        );

        // Set bind group and pipeline
        if let Some(ref bind_group) = self.bind_group {
            render_pass.set_bind_group(0, bind_group, &[]);
        }
        if let Some(ref pipeline) = self.pipeline {
            render_pass.set_pipeline(pipeline);
        }
        if let Some(ref index) = self.index {
            render_pass.set_index_buffer(index.buffer.slice(..), wgpu::IndexFormat::Uint32);
            render_pass.draw_indexed(0..index.size as u32, 0, 0..1);
        }
    }
}

// Helper extension to make Option handling easier
impl IndexBinding {
    fn take(&mut self) -> Option<IndexBinding> {
        // This would need manual implementation
        None
    }
}
